src/dev/lupio/lupio_tty.cc

```diff
@@ -0,0 +1,201 @@
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dev/lupio/lupio_tty.hh"
+
+#include "base/trace.hh"
+#include "debug/LupioTTY.hh"
+#include "dev/platform.hh"
+#include "mem/packet_access.hh"
+#include "params/LupioTTY.hh"
+
+#define LUPIO_TTY_INVAL    0x80000000
+
+/* Same fields for CTRL and STAT registers */
+#define LUPIO_TTY_WBIT  (1 << 0)
+#define LUPIO_TTY_RBIT  (1 << 1)
+
+namespace gem5
+{
+
+LupioTTY::LupioTTY(const Params &params)
+    : BasicPioDevice(params, params.pio_size),
+    writChar(-1), readChar(-1), writIntrEn(false), readIntrEn(false),
+    device(params.device), platform(params.platform)
+{
+    // setup serial device callbacks
+    device->regInterfaceCallback([this]() { dataAvailable(); });
+}
+
+void
+LupioTTY::dataAvailable()
+{
+      // unsure about these checks
+      // if interrupt enable bit is set and the read register is
+      // empty, then can raise an interrupt to receive data
+      if (readIntrEn && readChar == -1) {
+          DPRINTF(LupioTTY, "Data available, interrupting\n");
+          platform->postConsoleInt();
+     }
+}
+
+/*
+ * IRQ management
+ */
+void
+LupioTTY::lupioTTYUpdateIRQ()
+{
+    unsigned int irq;
+
+    irq = (writIntrEn && writChar != -1)
+        || (readIntrEn && readChar != -1);
+    if (irq) {
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, interrupting\n");
+        platform->postConsoleInt();
+    } else {
+//        platform->clearConsoleInt();
+        DPRINTF(LupioTTY, "LupioTTY InterEvent, not interrupting\n");
+    }
+}
+
+uint64_t
+LupioTTY::lupioTTYRead(uint8_t addr)
+{
+    uint32_t r = LUPIO_TTY_INVAL;
+
+    if (device->dataAvailable()) {
+        /* Drop characters */
+        if (readChar != -1)
+            fprintf(stderr, "Dropping characters\n");
+        else {
+            // read data from the terminal
+            readChar = device->readData();
+            lupioTTYUpdateIRQ();
+        }
+    }
+
+    switch (addr >> 2) {
+        case LUPIO_TTY_WRIT:
+            if (writChar != -1) {
+                DPRINTF(LupioTTY, "Accessing write register\n");
+                r = writChar;
+                writChar = -1;
+                lupioTTYUpdateIRQ();
+//                platform->clearConsoleInt();
+            }
+            break;
+        case LUPIO_TTY_READ:
+            if (readChar != -1) {
+                /* Return new character */
+                r = readChar;
+                readChar = -1;
+                lupioTTYUpdateIRQ();
+            }
+            break;
+        case LUPIO_TTY_CTRL:
+            DPRINTF(LupioTTY, "Accessing ctrl register\n");
+                        r = 0;
+                        if (writIntrEn)
+                                r |= LUPIO_TTY_WBIT;
+                        if (readIntrEn)
+                                r |= LUPIO_TTY_RBIT;
+            break;
```

Comment:
Indentation.

Reply:
And curly braces... I wish I could change our style to *require* them. I'm happy to show you examples of how skipping curly braces caused future bugs :)

Reply:
Do you require braces for single statements as well?

Reply:
I personally strongly prefer it. Skipping the braces has caused hard to find bugs in the past.